File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    answer = 0
    
    def dfs(cur_k, cnt):
        nonlocal answer
        answer = max(answer, cnt)
        
        for i in range(len(dungeons)):
            if not visited[i] and cur_k >= dungeons[i][0]:
                visited[i] = True
                dfs(cur_k - dungeons[i][1], cnt+1)
                visited[i] = False
    
    dfs(k, 0)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int k, vector<vector<int>> dungeons) {
    int ans = 0;
    int n = dungeons.size();
    
    vector<int> p;
    for(int i = 0; i < n; i++) p.push_back(i);
    
    do {
        int cur_k = k;
        int cnt = 0;
        
        for(int i = 0; i < n; i++) {
            int idx = p[i];
            if(cur_k >= dungeons[idx][0] && cur_k - dungeons[idx][1] >= 0) {
                cur_k -= dungeons[idx][1];
                cnt ++;
            } else break;
        }
        ans = max(ans, cnt);
            
        } while(next_permutation(p.begin(), p.end()));
    
    return ans;
}